packages/runtime/runtime-utils/src/dataStoreRegistry.ts

```diff
@@ -0,0 +1,62 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import {
+    FluidDataStoreRegistryEntry,
+    IFluidDataStoreRegistry,
+    FluidDataStoreRegistry,
+    FluidDataStoreRegistryEntries,
+    IProvideFluidDataStoreRegistry,
+} from "@fluidframework/runtime-definitions";
+
+export function createDataStoreRegistry(arg: FluidDataStoreRegistry): IProvideFluidDataStoreRegistry {
+    // Ensure that FluidDataStoreRegistry type did not change from original
+    // We do casting below and need to ensure that only these two types can come in
+    const entries: FluidDataStoreRegistryEntries | IProvideFluidDataStoreRegistry = arg;
+
+    const registry = entries as IProvideFluidDataStoreRegistry;
+    if (registry.IFluidDataStoreRegistry !== undefined) {
+        return registry;
+    }
+    return new DataStoreRegistry(entries as FluidDataStoreRegistryEntries);
+}
+
+class DataStoreRegistry implements IFluidDataStoreRegistry {
+    private readonly map: Map<string, FluidDataStoreRegistryEntry | Promise<FluidDataStoreRegistryEntry>> = new Map();
+
+    public get IFluidDataStoreRegistry() { return this; }
+
+    constructor(namedEntries: FluidDataStoreRegistryEntries) {
+        for (const entry of namedEntries) {
+            if (Array.isArray(entry)) {
+                this.map.set(entry[0], entry[1]);
+            } else {
+                this.map.set(entry.IFluidDataStoreFactory.type, entry);
```

Comment:
is type a new requirement? We do not expect our components to export type or know their own type in their factory export, we leave it up to each individual container to determine these strings. This is to avoid bad patterns of assuming registry pathing in components and also to handle the (small but could happen) case in which components might export the same type string.

Reply:
No, type was there for quite some time. Note that it's not really used directly by runtime (request to registry defined type).
But all explicit creation paths (i.e. DataObjectFactory.create*() APIs) need to know its type to create an object and assume that's an actual type used. in the registry.
No changes here in this PR

Reply:
I think type as a requirement is fine, but has it always been on IFluidDataStoreFactory (i.e. sourced from the component export)? Or is this some object that the container puts together?

Reply:
I think it all depends on who's loading package and what such code does with it
1. For example, wrapWithRuntimeFactoryIfNeeded() uses package name as a type for default data store factory.
2. Spaces, Vltava example uses type from fluidExport
3. examples\data-objects\shared-text will give its own names (mostly due to delay loading of factories). Similar for LocalCodeLoader - it just assign type = "default" and ignores whatever iis in fludExport/.

Personally, I'd love to see a world where components have hard-coded unique IDs similar to COM, but readable to the user. Something like <guid>-Scriptor[-<version>] But this is a bit out of scope of this PR IMHO.

Reply:
Yeah, in designs today/for the near future I steer away from hardcoded id strings from the component since it's of no real use to components. If they want to create copies of themselves they will have to use this.context.packagePath anyways to get their real pathing since they could be in a subregistry etc. Basically I've moved all of our strings to the container code only and do not let components touch it and I'd like to make sure that "wall" can still be in place.

Reply:
I think we should move this (and some other) discussions out of this PR. This PR does not change fundamentally how these mechanisms work. The only thing it does is
- consolidate runtime layers around existing interfaces, not around any particular way to represent data
- same for Aqueduct, but with a caveat - that interface is passed in for registries, otherwise it continues to do what it did before.
- adds helper methods to accept data in many formats to wrap it (on as needed bases) and expose interfaces that runtime expects.

I'll just point out that using this.context.packagePath has exactly same limitations as using type on factory itself. That is, the system expects only one entry with same name / factory instance in same registry node. Once it's unique, it really does not matter if we use name to reference it, or use factory instance to find its name from registry.
And in order to use this.context.packagePath, one needs to know it creates "child" of that node, "peer" of that node, or maybe root element (i.e. path in hands is for container registry).
It used to be that DataObjeect (actually code was in DataStoreRuntime - even worse) would do guessing game - it would guess - maybe I'm asked to created a child, maybe peer, maybe root element. All of that is gone - PureDataObjectFactory has dedicated createChildInstance / createPeerInstance / createInstance methods to clearly spell out intent.

Reply:
this.context.packagePath is much more reliable since it accounts for subregistry pathing and potential container renaming of keys to avoid collisions. There is no guarantee that the factory type string will be the one the container ultimately uses.

As long as it's not a hard requirement that components export this string and we can still control the usage from the container code then it should be fine either way, I just wanted to make sure that was still possible.